### PR TITLE
[DOCS] Update contributing.

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: docs/contributing.rst [hunk 1]
```diff
@@ -79,10 +80,10 @@ tests the root test directory has to be specified using the ``--testpath`` comma
 line option, e.g. ``./build/test/soltest -- --testpath ./test``.
 
 To disable the z3 tests, use ``./build/test/soltest -- --no-smt --testpath ./test`` and
```
Comment (ekpyron) [inline docs/contributing.rst]: It may be fine to ignore this, but I think this command will probably not work (it neither has ``--ipcpath``, nor ``--no-ipc``, so it'll fail). Maybe better to first document ``--no-ipc`` and then to add ``--no-smt`` on top of that, so that every command can actually be run.

File: docs/contributing.rst [hunk 2]
```diff
@@ -115,10 +116,10 @@ Example: ``./test/libsolidity/syntaxTests/double_stateVariable_declaration.sol``
         uint128 variable;
     }
     // ----
-    // DeclarationError: Identifier already declared.
+    // DeclarationError: (36-52): Identifier already declared.
 
-A syntax test must contain at least the contract under test itself, followed by the separator ``----``. The additional comments above are used to describe the
-expected compiler errors or warnings. This section can be empty in case that the contract should compile without any errors or warnings.
+A syntax test must contain at least the contract under test itself, followed by the separator ``// ----``. The following comments are used to describe the
```
Comment (ekpyron) [inline docs/contributing.rst:127]: Well, strictly speaking currently the separator is not necessary if the expected result is "no errors and no warnings" - in this case one can still use it, but it's optional - nevertheless I'd just keep it like this in the documentation - telling people to put the separator in there, even in this case, doesn't hurt.

File: docs/contributing.rst [hunk 3]
```diff
@@ -144,13 +145,13 @@ Running ``./test/isoltest`` again will result in a test failure:
...
         Obtained result:
             Success
 
 
-which prints the expected result next to the obtained result, but also provides a way to change edit / update / skip the current contract or to even quit.
-``isoltest`` offers several options for failing tests:
+``isoltest`` prints the expected result next to the obtained result, but also provides a way to change edit / update / skip the current contract or to even quit.
+It offers several options for failing tests:
 
 - edit: ``isoltest`` will try to open the editor that was specified before using ``isoltest --editor /path/to/editor``. If no path was set, this will result in a runtime error. In case an editor was specified, this will open it such that the contract can be adjusted.
```
Comment (ekpyron) [inline docs/contributing.rst]: We might want to add that - if the ``--editor`` option is not used -  ``isoltest`` will use the ``EDITOR`` environment variable, if set, and otherwise will try ``/usr/bin/editor``, if it exists (see https://github.com/ethereum/solidity/blob/b08d49338e2472ea579cbf8fcfb06a31bf517c1e/test/tools/isoltest.cpp#L283)).

File: docs/contributing.rst [hunk 4]
```diff
@@ -177,7 +178,8 @@ and re-run the test. It will now pass again:
 .. note::
 
     Please choose a name for the contract file, that is self-explainatory in the sense of what is been tested, e.g. ``double_variable_declaration.sol``.
```
Comment (ekpyron) [inline docs/contributing.rst]: Two typos: ``self-explainatory`` => ``self-explanatory`` and ``in the sense of what is *been* tested``.
Either ``of what is being tested`` or ``of what is to be tested``. Or just ``that is self-explanatory.``

File: docs/contributing.rst [hunk 5]
```diff
@@ -144,13 +145,13 @@ Running ``./test/isoltest`` again will result in a test failure:
             }
 
         Expected result:
-            DeclarationError: Identifier already declared.
+            DeclarationError: (36-52): Identifier already declared.
```
Comment (ekpyron) [inline docs/contributing.rst:156]: Maybe we should mention somewhere that ``(36-52)`` are source locations? Maybe even that ``isoltest`` will try to highlight them to make it easier to verify that they are correct?

File: docs/contributing.rst (hunk 4 above)
Comment (ekpyron) [reply docs/contributing.rst]: Or maybe even ``Please choose a name for the contract file, that explains what is being tested.`` or ``that explains what it tests.``.

File: docs/contributing.rst [hunk 6]
```diff
@@ -69,41 +70,46 @@ Thank you for your help!
...
+Solidity includes different types of tests. Most of them are bundled in the application
+called ``soltest``. Some of them require the ``aleth`` client in testing mode,
 some others require ``libz3`` to be installed.
 
-``soltest`` reads test contracts that are annotated with expected results
-stored in ``./test/libsolidity/syntaxTests``. In order for soltest to find these
-tests the root test directory has to be specified using the ``--testpath`` command
-line option, e.g. ``./build/test/soltest -- --testpath ./test``.
+To run a basic set of tests that neither require ``aleth`` nor ``libz3``, run
+``./scripts/soltest.sh --no-ipc --no-smt``. This script will run ``build/test/soltest``
```
Comment (leonardoalt) [inline docs/contributing.rst:82]: Won't run without `--testpath <path>`

File: docs/contributing.rst [hunk 7]
```diff
@@ -69,41 +70,46 @@ Thank you for your help!
...
 
-To disable the z3 tests, use ``./build/test/soltest -- --no-smt --testpath ./test`` and
-to run a subset of the tests that do not require ``cpp-ethereum``, use
-``./build/test/soltest -- --no-ipc --testpath ./test``.
+The option ``--no-smt`` disables the tests that require ``libz3`` and
+``--no-ipc`` disables those that requir ``aleth``.
 
-For all other tests, you need to install `cpp-ethereum <https://github.com/ethereum/cpp-ethereum/releases/download/solidityTester/eth>`_ and run it in testing mode: ``eth --test -d /tmp/testeth``.
+If you want to run the ipc tests (those test the semantics of the generated code),
+you need to install `aleth <https://github.com/ethereum/cpp-ethereum/releases/download/solidityTester/eth>`_ and run it in testing mode: ``aleth --test -d /tmp/testeth``.
```
Comment (axic) [inline docs/contributing.rst]: The file is called `eth`, referring to it as `aleth` is misleading. Once we move over to using aleth releases this would be simple.

File: docs/contributing.rst [hunk 8]
```diff
@@ -69,41 +70,46 @@ Thank you for your help!
...
-    While any version of ``cpp-ethereum`` should be usable, this cannot be guaranteed, and it is suggested to use the same version that is used by the Solidity continuous integration tests.
-    Currently the CI uses ``d661ac4fec0aeffbedcdc195f67f5ded0c798278`` of ``cpp-ethereum``.
+    Some versions of ``aleth`` cannot be used for testing. We suggest using the same version that is used by the Solidity continuous integration tests.
+    Currently the CI uses ``d661ac4fec0aeffbedcdc195f67f5ded0c798278`` of ``aleth``.
 
 Writing and running syntax tests
 --------------------------------
 
-As mentioned above, syntax tests are stored in individual contracts. These files must contain annotations, stating the expected result(s) of the respective test.
+Syntax tests check that the compiler generates the correct error messages for invalid code
```
Comment (axic) [inline docs/contributing.rst:110]: This should mention the directory they are placed in.

File: docs/contributing.rst (hunk 7 above)
Comment (axic) [reply docs/contributing.rst]: Or just rename the file on the release page.

File: docs/contributing.rst (hunk 6 above)
Comment (chriseth) [reply docs/contributing.rst:82]: `soltest.sh` auto-discovers the test path

Comment (leonardoalt) [reply docs/contributing.rst:82]: Ah ok

File: docs/contributing.rst [hunk 9]
```diff
@@ -69,41 +70,46 @@ Thank you for your help!
...
-line option, e.g. ``./build/test/soltest -- --testpath ./test``.
+To run a basic set of tests that neither require ``aleth`` nor ``libz3``, run
+``./scripts/soltest.sh --no-ipc --no-smt``. This script will run ``build/test/soltest``
+internally.
 
-To disable the z3 tests, use ``./build/test/soltest -- --no-smt --testpath ./test`` and
-to run a subset of the tests that do not require ``cpp-ethereum``, use
-``./build/test/soltest -- --no-ipc --testpath ./test``.
+The option ``--no-smt`` disables the tests that require ``libz3`` and
+``--no-ipc`` disables those that requir ``aleth``.
```
Comment (erak) [inline docs/contributing.rst]: Typo: "require"

File: docs/contributing.rst (hunk 7 above)
Comment (erak) [reply docs/contributing.rst]: I'd would split this up and mention that you can either download the executable (``eth``) or build it locally (``aleth``).

Comment (erak) [reply docs/contributing.rst]: Wait, is it a Circle build? We could get a new one and replace ``eth``.

Comment (axic) [reply docs/contributing.rst]: @ekpyron tried that but there are some issues

Comment (chriseth) [reply docs/contributing.rst]: Actually if we recommend to use the CI binary some lines down, we should also link that here.

File: docs/contributing.rst [hunk 10]
```diff
@@ -144,15 +152,15 @@ Running ``./test/isoltest`` again will result in a test failure:
...
             Success
 
 
-which prints the expected result next to the obtained result, but also provides a way to change edit / update / skip the current contract or to even quit.
-``isoltest`` offers several options for failing tests:
+``isoltest`` prints the expected result next to the obtained result, but also provides a way to change edit / update / skip the current contract or to even quit.
+It offers several options for failing tests:
 
-- edit: ``isoltest`` will try to open the editor that was specified before using ``isoltest --editor /path/to/editor``. If no path was set, this will result in a runtime error. In case an editor was specified, this will open it such that the contract can be adjusted.
+- edit: ``isoltest`` will try to open the contract in an editor so you can adjust it. It will either use the editor given on the command line (as ``isoltest --editor /path/to/editor``), in the environment variable ``EDITOR`` or just ``/usr/bin/editor`` (in this order).
```
Comment (erak) [inline docs/contributing.rst]: I'd remove the future tense here.

File: docs/contributing.rst [hunk 11]
```diff
@@ -115,10 +122,11 @@ Example: ``./test/libsolidity/syntaxTests/double_stateVariable_declaration.sol``
...
     }
     // ----
-    // DeclarationError: Identifier already declared.
+    // DeclarationError: (36-52): Identifier already declared.
 
-A syntax test must contain at least the contract under test itself, followed by the separator ``----``. The additional comments above are used to describe the
-expected compiler errors or warnings. This section can be empty in case that the contract should compile without any errors or warnings.
+A syntax test must contain at least the contract under test itself, followed by the separator ``// ----``. The following comments are used to describe the
+expected compiler errors or warnings. The number range denotes the location in the source where the error occurred.
+The section after the separator can be empty in case the contract should compile without any errors or warnings.
```
Comment (erak) [inline docs/contributing.rst]: Maybe rephrase: "In case the contract should compile without any errors or warning, the separator can be left out and the section after it can be empty" ? I'd somehow take https://github.com/ethereum/solidity/pull/5084/files#r220308437 into account. Perhaps add an example of such a contract?

File: docs/contributing.rst [hunk 12]
```diff
@@ -69,41 +70,47 @@ Thank you for your help!
...
+internally.
 
-To disable the z3 tests, use ``./build/test/soltest -- --no-smt --testpath ./test`` and
-to run a subset of the tests that do not require ``cpp-ethereum``, use
-``./build/test/soltest -- --no-ipc --testpath ./test``.
+The option ``--no-smt`` disables the tests that require ``libz3`` and
+``--no-ipc`` disables those that require ``aleth``.
 
-For all other tests, you need to install `cpp-ethereum <https://github.com/ethereum/cpp-ethereum/releases/download/solidityTester/eth>`_ and run it in testing mode: ``eth --test -d /tmp/testeth``.
+If you want to run the ipc tests (those test the semantics of the generated code),
```
Comment (erak) [inline docs/contributing.rst:88]: I'm still thinking about splitting this section up into platform-dependent download / build instructions. We could make a Linux-only section out of this and create a cross-platform one where we'd link the aleth build instructions.

Comment (chriseth) [reply docs/contributing.rst:88]: I guess this already is rather linux-specific.

**Review by erak — APPROVED** [review]
